src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -1218,71 +1252,132 @@ void ObjectMonitor::VThreadEpilog(JavaThread* current, ObjectWaiter* node) {
   }
 }
 
-// By convention we unlink a contending thread from EntryList|cxq immediately
+// Return the tail of the _entry_list. If the tail is currently not
+// known, find it by walking from the head of _entry_list, and while
+// doing so assign the _prev pointers to create a doubly linked list.
+ObjectWaiter* ObjectMonitor::entry_list_tail(JavaThread* current) {
+  assert(has_owner(current), "invariant");
+  ObjectWaiter* w = _entry_list_tail;
+  if (w != nullptr) {
+    return w;
+  }
+  // Need acquire here to match the implicit release of the cmpxchg
+  // that updated _entry_list, so we can access w->_next.
+  w = Atomic::load_acquire(&_entry_list);
+  assert(w != nullptr, "invariant");
+  if (w->next() == nullptr) {
+    _entry_list_tail = w;
+    return w;
+  }
+  ObjectWaiter* prev = nullptr;
+  while (w != nullptr) {
+    assert(w->TState == ObjectWaiter::TS_ENTER, "invariant");
+    w->_prev = prev;
+    prev = w;
+    w = w->next();
+  }
+  _entry_list_tail = prev;
+  return prev;
+}
+
+// By convention we unlink a contending thread from _entry_list immediately
 // after the thread acquires the lock in ::enter().  Equally, we could defer
 // unlinking the thread until ::exit()-time.
```

Comment:
Since you're here, remove these two lines 1222-1223.  I really don't think pointing out an alternate implementation that we did not choose is helpful to understanding this code.